### DISCUSSION TITLE
Dynamic Simulation of Heat Exchanger

### DISCUSSION THREAD

**Discussion by Simon20009** [IDAES/idaes-pse#1496] [Q&A]
Hi

I am trying to create a dynamic simulation with a heat exchanger (which is to be integrated into a heat pump circuit), but I encounter problems when initializing or solving.

Code for 1D heat exchanger
[Wärmetauscher1D.ipynb.txt](https://github.com/user-attachments/files/17307846/Warmetauscher1D.ipynb.txt)
[Propane.py.txt](https://github.com/user-attachments/files/17307847/Propane.py.txt)

The model can be solved stationary.

In the dynamic model, I still have 2 degrees of freedom open at the end, but I don't know which ones are required.

Can you tell me what my mistake is?

Thanks for your help.

**Errors:**
**Initializing:**

```
2024-10-09 13:58:33 [INFO] idaes.init.fs.heat_exchanger.hot_side: Initialization Complete
2024-10-09 13:58:33 [INFO] idaes.init.fs.heat_exchanger.cold_side: Initialization Complete
---------------------------------------------------------------------------
InitializationError                       Traceback (most recent call last)
Cell In[10], line 1
----> 1 m.fs.heat_exchanger.initialize(outlvl=idaeslog.INFO)

File ~\.conda\envs\IDAES\lib\site-packages\idaes\core\base\unit_model.py:540, in UnitModelBlockData.initialize(blk, *args, **kwargs)
    537     c.deactivate()
    539 # Remember to collect flags for fixed vars
--> 540 flags = blk.initialize_build(*args, **kwargs)
    542 # If costing block exists, activate and initialize
    543 for c in init_order:

File ~\.conda\envs\IDAES\lib\site-packages\idaes\models\unit_models\heat_exchanger_1D.py:862, in HeatExchanger1DData.initialize_build(self, hot_side_state_args, cold_side_state_args, outlvl, solver, optarg, duty)
    859 self.cold_side.release_state(flags_cold_side)
    861 if res is not None and not check_optimal_termination(res):
--> 862     raise InitializationError(
    863         f"{self.name} failed to initialize successfully. Please check "
    864         f"the output logs for more information."
    865     )
    867 init_log.info("Initialization Complete.")

InitializationError: fs.heat_exchanger failed to initialize successfully. Please check the output logs for more information.
```

**Solver:**
```
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Called fg_read, err: 0 (0 is good)
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: ---------------------------------------------------
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: DAE: 0
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Reading nl file: C:\Users\SIMON~1.END\AppData\Local\Temp\tmp2adhab7u.pyomo.nl
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of constraints: 451
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of nonlinear constraints: 160
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of linear constraints: 291
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of inequalities: 0
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of variables: 453
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of integers: 0
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of binary: 0
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of objectives: 0 (Ignoring)
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of non-zeros in Jacobian: 1214 
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: Number of degrees of freedom: 2
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: ---------------------------------------------------
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: ERROR: Degrees of freedom not equal to 0
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: ERROR: Solver (asl) returned non-zero return code (2)
2024-10-09 14:04:58 [INFO] idaes.solve.petsc-dae: ERROR: See the solver log above for diagnostic information.
---------------------------------------------------------------------------
ApplicationError                          Traceback (most recent call last)
Cell In[11], line 3
      1 from idaes.core.solvers import petsc
----> 3 result = petsc.petsc_dae_by_time_element(
      4     m,
      5     time=m.fs.time,
      6     ts_options={
      7         "--ts_type": "beuler",
      8         "--ts_dt": 0.1,
      9         "--ts_monitor": "",  # set initial step to 0.1
     10         "--ts_save_trajectory": 1,
     11     },
     12 )
     13 tj = result.trajectory  # trajectroy data
     14 res = result.results

File ~\.conda\envs\IDAES\lib\site-packages\idaes\core\solvers\petsc.py:592, in petsc_dae_by_time_element(m, time, timevar, initial_constraints, initial_variables, detect_initial, skip_initial, initial_solver, initial_solver_options, ts_options, keepfiles, symbolic_solver_labels, between, interpolate, calculate_derivatives, previous_trajectory, representative_time, snes_options)
    590             _sub_problem_scaling_suffix(m, t_block)
    591             with idaeslog.solver_log(solve_log, idaeslog.INFO) as slc:
--> 592                 res = initial_solver_obj.solve(
    593                     t_block,
    594                     tee=slc.tee,
    595                     symbolic_solver_labels=symbolic_solver_labels,
    596                 )
    597             res_list.append(res)
    599 tprev = t0

File ~\.conda\envs\IDAES\lib\site-packages\pyomo\opt\base\solvers.py:628, in OptSolver.solve(self, *args, **kwds)
    626     elif hasattr(_status, 'log') and _status.log:
    627         logger.error("Solver log:\n" + str(_status.log))
--> 628     raise ApplicationError("Solver (%s) did not exit normally" % self.name)
    629 solve_completion_time = time.time()
    630 if self._report_timing:

ApplicationError: Solver (asl) did not exit normally
```

**Comment by andrewlee94** [comment]
@Simon20009 The short answer is that when you make a model dynamic you add extra degrees of freedom. You also need to think carefully about what it is you are trying to model and what matters in the system.

For example, in heater and heat exchanges the thing that matters most is the thermal mass of the metal wall, not the material (and energy) holdup of the fluids. Thus, in many cases the thing you want to do is add a dynamic energy holdup for the wall material, but leave the fluid flows as (quasi)-steady-state.  You might want to look at the lumped capacitance heat exchanger model (https://github.com/IDAES/idaes-pse/blob/main/idaes/models/unit_models/heat_exchanger_lc.py), which is generally a good place to start for dynamic heat exchangers.

**Reply by Simon20009** [reply]
@andrewlee94  Thank you very much for the quick response.

So if I understand correctly, a “dynamic” simulation only works with the lumped capacitance heat exchanger, because I can't add any heat capacity to the others heat exchanger models without modifying the code , or?

I have looked at the example (https://idaes-pse.readthedocs.io/en/latest/reference_guides/model_libraries/generic/unit_models/heat_exchanger_lc.html) and will carry out various tests in the near future.

**Reply by andrewlee94** [reply]
No, that is not correct. Almost any model in IDAES can handle dynamics, however in all cases the user needs to think carefully about what it means for a model to be dynamic and to provide additional equations to describe how the system operates under dynamic changes.

Dynamic modeling is far more than just adding an accumulation term to the material, energy and momentum balances; if you sit down and write out the equations for a steady-state model and a dynamic model you will find the degrees of freedom are different. In general, dynamic models have one additional degree of freedom (at least) than an equivalent steady-state model. This degree of freedom is due to the fact that you need to describe the relationship between accumulation in the unit and the flow leaving the unit; one common (but not the only) way to describe this is via a pressure driven flow relationship (i.e. flow is somehow related to pressure). However, there are many ways to do this, and which is best depends on exactly what system you are modeling, thus that is left to the user to provide.

Equally important however is to realise that not all dynamic behaviors are equal - consider the rate of propagation of a pressure wave (speed of sound) versus accumulation of a component in a mixing tank. The pressure will generally equilibrate nearly instantaneously to any change, whilst the concentration of the species in the tank will take much longer (generally >5x the residence time of the tank). Thus, you need to think about which dynamics (time constants) are actually important - mixing time scales of different magnitude in a single model introduces "stiffness", meaning you will not be able to solve the model.

For the case of a heater/heat exchanger - pressure waves generally move throguh the system so fast that you should assume quasi-steady-state, and the fluid residence times are generally less than a minute (at most), whilst the time for the wall to reach thermal equilibrium is much longer. Thus, in most (but not all) cases, when modeling dynamic heat exchangers you really only want to consider the dynamics of the wall and assume the rest of the system is at quasi-steady-state. You can try to model other dynamics, but you will generally find the system very quickly becomes stiff and will be difficult or impossible to solve (and if it does solve, you will find that the fluid phases were close to quasi-steady-state anyway).
